### PR TITLE
Run the UTs for PRs only with the oldest language version

### PR DESCRIPTION
**Pull request by dmitry-golovinov-sonarsource**

### CODE REVIEW

File: scripts/build/dev-build.ps1 [hunk 1]
```diff
@@ -86,7 +86,7 @@ try {
         # It considers to run under CI/CD only if the following environment variable is set.
         # When running this script, we want to run all unit tests.
         # see analyzers\tests\SonarAnalyzer.UnitTest\TestFramework\ParseOptionsHelper.cs
-        $env:SYSTEM_DEFINITIONID="set"
+        $env:Build.Reason="set"
```
Comment (pavel-mikula-sonarsource) [inline scripts/build/dev-build.ps1:89]: Lines 85 and 86 above are slightly out of date

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    CS.LanguageVersion.CSharp6,
+                    CS.LanguageVersion.CSharp7,
+                    CS.LanguageVersion.CSharp7_1,
+                    CS.LanguageVersion.CSharp7_2,
+                    CS.LanguageVersion.CSharp7_3,
+                    CS.LanguageVersion.CSharp8,
+                    CS.LanguageVersion.CSharp9
+                };
+                // the following assertion is expected to fail when we add the support of C#10
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs:49]: To make it simple and version-independent so we can only update the expectations
```suggestion
                // This should fail when we add new language version
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                VB.LanguageVersion[] vbVersions =
+                {
+                    VB.LanguageVersion.VisualBasic12,
+                    VB.LanguageVersion.VisualBasic14,
+                    VB.LanguageVersion.VisualBasic15,
+                    VB.LanguageVersion.VisualBasic15_3,
+                    VB.LanguageVersion.VisualBasic15_5,
+                    VB.LanguageVersion.VisualBasic16
+                };
+                ParseOptionsHelper.FromVisualBasic12.Should().Contain(vbVersions.Select(x => new VB.VisualBasicParseOptions(x)));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs:61]: Let's prepare this before the `if`
```
            var vbVersions = ParseOptionsHelper.FromVisualBasic12.Cast<VB.VisualBasicParseOptions>().Select(x => x.LanguageVersion);
```
and then just assert it in a simple way inside `if` and `else`. It should make it easier to read. Same for C#
```suggestion
            vbVersions.Should().BeEquivalentTo(
                    VB.LanguageVersion.VisualBasic12,
                    VB.LanguageVersion.VisualBasic14,
                    VB.LanguageVersion.VisualBasic15,
                    VB.LanguageVersion.VisualBasic15_3,
                    VB.LanguageVersion.VisualBasic15_5,
                    VB.LanguageVersion.VisualBasic16);

```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            }
+            else
+            {
+                ParseOptionsHelper.FromCSharp6.Should().Contain(new CS.CSharpParseOptions(CS.LanguageVersion.CSharp6));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs:65]: The `.Contain` doesn't care if there are more items in the list AFAICT.
```suggestion
                vbVersions.Should().HaveCount(1).And.Contain(VB.LanguageVersion.VisualBasic12);
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestContextHelper.cs [hunk 5]
```diff
@@ -25,6 +25,9 @@ namespace SonarAnalyzer.UnitTest.Helpers
     internal static class TestContextHelper
     {
         public static bool IsAzureDevOpsContext =>
-            Environment.GetEnvironmentVariable("SYSTEM_DEFINITIONID") != null;
+            Environment.GetEnvironmentVariable("Build.Reason") != null;
+
+        public static bool IsNotPullRequestBuild =>
+            Environment.GetEnvironmentVariable("Build.Reason") != "PullRequest";
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestContextHelper.cs:31]: Let's have positive state
```suggestion
        public static bool IsPullRequestBuild =>
            Environment.GetEnvironmentVariable("Build.Reason") == "PullRequest";
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelper.cs [hunk 6]
```diff
@@ -94,7 +94,7 @@ language switch
             };
 
         private static ImmutableArray<ParseOptions> FilterByEnvironment(this IEnumerable<ParseOptions> options) =>
-            TestContextHelper.IsAzureDevOpsContext
+            TestContextHelper.IsNotPullRequestBuild
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelper.cs:97]: When reading this, I miss the `local build uses single version as well` part. Wasn't the `IsAzureDevOpsContext` meant to be preserved here?

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs (hunk 4 above)
Comment (dmitry-golovinov-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs:65]: Turned into _vbVersions.Should().BeEquivalentTo_

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelper.cs [hunk 7]
```diff
@@ -94,9 +94,9 @@ language switch
             };
 
         private static ImmutableArray<ParseOptions> FilterByEnvironment(this IEnumerable<ParseOptions> options) =>
-            TestContextHelper.IsAzureDevOpsContext
-                ? options.ToImmutableArray()
-                : ImmutableArray.Create(options.First()); // Use only the oldest version for local test run and debug
+            TestContextHelper.IsLocalOrPullRequestBuild
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelper.cs:97]: _Open question:_ This works, but wouldn't this be better? To keep the helper functions atomic.
```suggestion
            TestContextHelper.IsAzureDevOpsContext && !TestContextHelper.IsPullRequestBuild
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelperTest.cs]: Same for C#, please.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**